README.md
![workflow](https://github.com/addicteduser/dnd-hp-calc-discordbot/actions/workflows/ci.yml/badge.svg?branch=master)

# Valron the HP Calculator

![Valron the HP Calculator](https://i.imgur.com/0bByXQ4.png?s=200)

_Art by [Zentrend](https://www.instagram.com/neil_is_zen/)._

[![Discord Bots](https://top.gg/api/widget/666625461811413008.svg)](https://top.gg/bot/666625461811413008)

Discord Bot for computing your AL D&amp;D 5e character's hit points, given the Constitution modifier, its classes and levels, and other HP modifiers such as Tough feat or being a Hill Dwarf.

## Discord Server Invitation

Invite it to your Discord server with this [link](https://discordapp.com/api/oauth2/authorize?client_id=666625461811413008&permissions=11264&scope=bot).

## Usage

### Command

`/hp <con_modifier> <classA#/classB#/etc> [hp_mod1/hp_mod2/etc]`

#### Example

`/hp 3 fighter1/barb2/paladin1`

#### Example with HP modifiers

`/hp 3 fighter1/barb2/paladin1 tough/hilldwarf`

### Other commands

- `/help` - main help command
- `/options` - to see the list of supported classes and HP modifiers
- `/links` - to view some helpful links

## Dependencies

- [Python 3.8](https://docs.python.org/3.8/)
- [disnake.py v2.5.2](https://docs.disnake.dev/en/v2.5.2/), a modern, easy to use, feature-rich, and async ready API wrapper for Discord
- See `requirements.txt` for full list of dependencies
- For unit tests (see `requirements-test.txt`):
  - [pytest](https://docs.pytest.org/en/stable/) for running unit tests.
  - [pytest-asyncio](https://pypi.org/project/pytest-asyncio/) for testing asyncio code with pytest.

## Local Development

- Install dependencies with `pip install -r requirements.txt`
- Run with `python bot.py`
- Install unit test dependencies with `pip install -r requirements-test.txt`
- Run unit tests with `pytest`

## Dontation/Support

Hello, hello! This bot that I have made is just a little passion project of mine. Any support or donation ([PayPal](https://paypal.me/addicteduser) | [Ko-Fi](https://ko-fi.com/addicteduser) | [GCash](https://i.imgur.com/fnMORVa.jpg)) is much appreciated to keep up with the hosting fees. Thank you very much!

For feedback or suggestions, join the support discord server [here](https://discord.gg/waCBQuD).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--
## [Unreleased](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/LATEST_TAG...HEAD) - YYYY-MM-DD

### Added
- for new features.

### Changed
- for changes in existing functionality.

### Deprecated
- for soon-to-be removed features.

### Removed
- for now removed features.

### Fixed
- for any bug fixes.

### Security
- in case of vulnerabilities.

-->

## [Unreleased](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/3.0.0...HEAD) - YYYY-MM-DD

## [3.0.0](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/2.0.0...3.0.0) - 2024-12-21

### Changed

- Update disnake from 2.5.2 to 2.9.1.

## [2.0.0](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.6.2...2.0.0) - 2022-08-31

### Added

- Migrate to use [slash command (`/`)](https://discord.com/blog/welcome-to-the-new-era-of-discord-apps/).
- Normal bot commands also work when mentioning the bot.

### Removed

- Remove support for prefix commands.

## [1.6.2](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.6.1...1.6.2) - 2022-08-11

### Changed

- Send HP calculation reply inside an embed.
- Append reminder about [migrating to Slash Commands](https://discord.com/blog/welcome-to-the-new-era-of-discord-apps/) by August 31, 2022.

## [1.6.1](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.6.0...1.6.1) - 2021-10-27

### Added

- Add `test-requirements.txt` for listing unit test dependencies.
- Add unit tests on parsing user input.

### Changed

- Migrate CI from Travis CI to GitHub Actions.

### Removed

- Remove usage of privileged server members intent (see [this](https://discordpy.readthedocs.io/en/stable/intents.html) for more info).

### Fixed

- Fix linking to Ko-Fi and GCash.

### Security

- Removed some hardcoded info.

## [1.6.0](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.5.1...1.6.0) - 2021-03-10

### Added

- Add `CHANGELOG.md`.
- Add `.env.sample`.
- Add `.travis.yml` for continuous integration.

### Changed

- Implement pylint suggestions.
- Move utility/helper modules to `utils/` and refactor affected code.
- Move unit tests to `tests/`.
- Accessing Discord token is now in `.env` instead of `secrets.py`.
- Optimize and parametrize ([`@pytest.mark.parametrize`](https://docs.pytest.org/en/stable/parametrize.html)) test functions.
- Add Travis CI badge in `README.md`.
- Include Ko-Fi and GCash links in Donation/Support section of `README.md`.

## [1.5.1](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.5.0...1.5.1) - 2021-02-08

### Added

- Add a link to GCash QR code in `?links`.

## [1.5.0](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.4.0...1.5.0) - 2021-02-05

### Added

- Add unit test.

### Changed

- Internal code restructuring and optimizations.
- Update discord.py from 1.3.4 to 1.6.0.

## [1.4.0](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.3.1...1.4.0) - 2020-09-09

### Changed

- Internal code restructuring.
- Lessen "Valron is typing..." time.
- Include `pal` to the list of `paladin` shortcuts.
- Remove `pally` from the list of `paladin` shortcuts.

## [1.3.1](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.3.0...1.3.1) - 2020-08-31

### Added

- Add activity status.

### Changed

- Lessened "Valron is typing..." time.

## [1.3.0](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.2.0...1.3.0) - 2020-08-28

### Added

- Add support for `artificer` class.
- Implement embeds.
- Add `?options` and `?links` commands.
- Implement "Valron is typing..." :3

### Changed

- Update help command from `?hphelp` to `?help`.
- Update list of classes (and shortcuts, e.g. `cleric` can be `cl` or `c`).
- Internal code restructuring and optimizations.

## [1.2.0](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.1.0...1.2.0) - 2020-07-22

### Changed

- Update discord.py from 1.2.5 to 1.3.4.

## [1.1.0](https://github.com/addicteduser/dnd-hp-calc-discordbot/compare/1.1.0...master) - 2020-03-25

### Changed

- Change the command usage from `!hp` to `?hp`.


utils/constants.py
"""Constants

This module provides the constant values.
"""

# pylint: disable=C0103

from typing import List

from .classes import DndClass

SHOW_PERF = False


###################
## DATA BUILDERS ##
###################
def DND_CLASSES() -> List[DndClass]:
    """Returns the list of possible D&D classes.

    Returns
    -------
    List[DndClass]
        The list of possible D&D classes.
    """
    classes = []

    classes.append(DndClass("Artificer", ["artificer", "art", "a"], 8))
    classes.append(DndClass("Barbarian", ["barbarian", "barb", "bb"], 12))
    classes.append(DndClass("Bard", ["bard", "bd"], 8))
    classes.append(DndClass("Cleric", ["cleric", "cl", "c"], 8))
    classes.append(DndClass("Druid", ["druid", "dr", "d"], 8))
    classes.append(DndClass("Fighter", ["fighter", "fight", "f"], 10))
    classes.append(DndClass("Monk", ["monk", "mk", "m"], 8))
    classes.append(DndClass("Paladin", ["paladin", "pal", "p"], 10))
    classes.append(DndClass("Ranger", ["ranger", "ra"], 10))
    classes.append(DndClass("Rogue", ["rogue", "ro"], 8))
    classes.append(DndClass("Sorcerer", ["sorcerer", "sorc", "s"], 6))
    classes.append(
        DndClass(
            "Draconic Sorcerer",
            ["draconicsorcerer", "draconicsorc", "dracsorc", "ds"],
            6,
        )
    )
    classes.append(DndClass("Warlock", ["warlock", "lock", "wr"], 8))
    classes.append(DndClass("Wizard", ["wizard", "wiz", "wz"], 6))

    return classes


def DND_ALIASES() -> List[List[str]]:
    """Returns the list of alias lists per D&D class.

    Returns
    -------
    List[List[str]]
        The list of alias lists per D&D class.
    """
    aliases = []

    for dnd_class in DND_CLASSES():
        aliases.append(dnd_class.aliases)

    return aliases


###############
## CONSTANTS ##
###############
def HILLDWARF_MODS():
    """Returns the Hill Dwarf HP modifier aliases."""
    return ["hilldwarf", "hdwarf", "hd"]


def BERSERKER_AXE_MODS():
    """Returns the Berserker Axe HP modifier aliases."""
    return ["berserkeraxe", "axe", "ba"]


def TOUGH_MODS():
    """Returns the Tough Feat HP modifier aliases."""
    return ["tough", "t"]


def HP_MODS():
    """Returns the list of HP modifier aliases."""
    return HILLDWARF_MODS() + BERSERKER_AXE_MODS() + TOUGH_MODS()


def HP_MOD_ALIASES():
    """Returns the list of list of HP modifer aliases."""
    return [HILLDWARF_MODS(), BERSERKER_AXE_MODS(), TOUGH_MODS()]


###########
## LINKS ##
###########
TOP_GG_LINK = "https://top.gg/bot/666625461811413008"
GITHUB_LINK = "https://github.com/addicteduser/dnd-hp-calc-discordbot"
IMG_LINK = "https://i.imgur.com/0bByXQ4.png"

DISCORD_INVITE_LINK = (
    "https://discordapp.com/api/oauth2/authorize?"
    + "client_id=666625461811413008&"
    + "permissions=11264&scope=bot"
)
SUPPORT_SERVER_LINK = "https://discord.gg/waCBQuD"

PAYPAL_LINK = "https://paypal.me/addicteduser"
KOFI_LINK = "https://ko-fi.com/addicteduser"
GCASH_QR_CODE = "https://i.imgur.com/fnMORVa.jpg"
